File: apps/backend/app/api/v1/runtime_flags.py
"""Runtime feature flags — read-only.

A single GET endpoint that returns the current values of the runtime
flags Settings exposes. The frontend polls this on app boot (and on a
60s stale window via TanStack Query) so a flag flip on the server side
becomes visible without any client redeploy.

L20.5 ships the wire shape only — the response reads straight from
:mod:`app.core.config`. L21-Sec adds a Redis-backed override layer
(``runtime_flags:override:<key> -> 0/1``) so an admin can flip a flag
without restarting the API. The endpoint stays anon-readable: nothing
returned here is sensitive, and gating it on auth would force every
public page to know the user's session before it could decide which
code path to mount.
"""

from __future__ import annotations

from fastapi import APIRouter
from pydantic import BaseModel, ConfigDict

from app.core.config import get_settings

router = APIRouter()


class RuntimeFlags(BaseModel):
    """Public runtime-flags payload.

    Every flag here defaults to the value in :class:`Settings`. Adding a
    new flag is two lines (Settings + this schema). Removing or renaming
    one is a breaking change for any client cached behind a CDN — bump a
    version field if that ever becomes a problem; today there's nothing
    to migrate.
    """

    model_config = ConfigDict(from_attributes=False)

    # L21a/L21b — SSE streaming tutor turns. OFF until L21b's flag-flip.
    tutor_streaming: bool


@router.get(
    "/runtime-flags",
    response_model=RuntimeFlags,
    summary="Public runtime feature flags",
    tags=["runtime-flags"],
)
async def get_runtime_flags() -> RuntimeFlags:
    s = get_settings()
    return RuntimeFlags(
        tutor_streaming=s.feature_tutor_streaming,
    )

File: apps/frontend/src/lib/runtime-flags.ts
"use client";

import { useQuery } from "@tanstack/react-query";
import { RuntimeFlagsApi, type RuntimeFlags } from "@/lib/api/endpoints";
import { qk } from "@/lib/query/keys";

/**
 * L20.5 — Runtime feature flags.
 *
 * `useRuntimeFlags()` reads `/api/v1/runtime-flags` once per page,
 * caches for 60s, and re-validates on focus. Defaults to all-off
 * (the safe pre-flip state) while the request is in flight so a
 * page's first paint doesn't accidentally unlock a feature.
 *
 * The L21b flag-flip PR will toggle `tutor_streaming` on by writing
 * to the same Settings field (L21-Sec will additionally add a Redis
 * override layer so an admin can flip without a redeploy). Callers
 * gate streaming UI with `flags.tutor_streaming` so the codepaths
 * stay dormant until that flip.
 */

const DEFAULT_FLAGS: RuntimeFlags = {
  tutor_streaming: false,
};

export function useRuntimeFlags(): RuntimeFlags {
  const q = useQuery({
    queryKey: qk.runtimeFlags,
    queryFn: () => RuntimeFlagsApi.get(),
    staleTime: 60_000,
    gcTime: 5 * 60_000,
    refetchOnWindowFocus: true,
    retry: 1,
  });
  return q.data ?? DEFAULT_FLAGS;
}
